wal/src/message_queue_impl/region_meta.rs

```diff
@@ -0,0 +1,587 @@
+                latest_marked_deleted: inner.latest_marked_deleted,
+                current_high_watermark: inner.current_high_watermark,
+                safe_delete_offset: None,
+            }
+        } else {
+            let offset = inner
+                .start_sequence_offset_mapping
+                .get(&inner.latest_marked_deleted);
+            assert!(
+                inner.next_sequence_num > inner.latest_marked_deleted,
+                "unexpected state, next sequence should be greater than latest marked deleted
+                while has new writing after init/flush, but now are {} and {}",
+                inner.next_sequence_num,
+                inner.latest_marked_deleted
+            );
+            assert!(
+                offset.is_some(),
+                "unexpected state, offset should be found while has new writing after init/flush
+                but now next sequence num:{}, latest marked deleted:{}, mapping:{:?}",
+                inner.next_sequence_num,
+                inner.latest_marked_deleted,
+                inner.start_sequence_offset_mapping
+            );
+
+            TableMetaData {
+                table_id: self.table_id,
+                next_sequence_num: inner.next_sequence_num,
+                latest_marked_deleted: inner.latest_marked_deleted,
+                current_high_watermark: inner.current_high_watermark,
+                safe_delete_offset: offset.copied(),
+            }
+        }
+    }
+}
+
+// TODO: will be made use of later.
+#[allow(unused)]
+/// Table meta data, will be updated atomically by mutex.
+#[derive(Debug, Default)]
+struct TableMetaInner {
+    /// Next sequence number for the new log.
+    ///
+    /// It will be updated while having pushed logs successfully.
+    next_sequence_num: SequenceNumber,
+
+    /// The lasted marked deleted sequence number. The log with
+    /// a sequence number smaller than it can be deleted safely.
+    ///
+    /// It will be updated while having flushed successfully.
+    latest_marked_deleted: SequenceNumber,
+
+    /// The high watermark after this table's latest writing.
+    current_high_watermark: Offset,
+
+    /// Map the start sequence number to start offset in every write.
+    ///
+    /// It will be removed to the mark deleted sequence number after flushing.
+    start_sequence_offset_mapping: BTreeMap<SequenceNumber, Offset>,
+}
+
+// TODO: will be made use of later.
+#[allow(unused)]
+#[derive(Debug, Default)]
+pub struct TableMetaData {
+    pub table_id: TableId,
+    pub next_sequence_num: SequenceNumber,
+    pub latest_marked_deleted: SequenceNumber,
+    pub current_high_watermark: Offset,
+    pub safe_delete_offset: Option<Offset>,
+}
+
+/// Message queue implementation's meta value.
+///
+/// Include all tables(of current shard) and their next sequence number.
+pub struct RegionMetaSnapshot {
+    pub entries: Vec<TableMetaData>,
+}
+
+pub struct OffsetRange {
```

Comment:
Add comments to explain that the end offset is inclusive. BTW, it can't express an empty range if we make the end offset inclusive, and is that OK?